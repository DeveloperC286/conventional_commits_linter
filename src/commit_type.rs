use clap::ValueEnum;

#[derive(Clone, Debug, PartialEq, ValueEnum)]
pub enum CommitType {
    // Allow the Conventional Commits type to be any value.
    Any,
    // Allow the Conventional Commits type to only be (`build`, `chore`, `ci`, `docs`, `feat`, `fix`, `perf`, `refactor`, `style`, `test`, `revert`).
    Angular,
}
