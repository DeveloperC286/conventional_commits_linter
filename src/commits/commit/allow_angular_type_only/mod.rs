use regex::Regex;

use super::*;

pub(super) fn lint(commit_message: &str) -> Result<(), LintingError> {
    lazy_static! {
        static ref ANGULAR_TYPE_REGEX: Regex = Regex::new(&format!(
            r"(?i)^{OPTIONAL_PRECEDING_WHITESPACE}{ANGULAR_TYPE}{OPTIONAL_EXCLAMATION}{OPTIONAL_EMPTY_SCOPE_OR_SCOPE}{OPTIONAL_EXCLAMATION}:",
        ))
        .unwrap();
    }

    match ANGULAR_TYPE_REGEX.is_match(commit_message) {
        true => Ok(()),
        false => Err(LintingError::NonAngularType),
    }
}
