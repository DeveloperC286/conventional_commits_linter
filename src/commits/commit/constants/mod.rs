pub(super) const PRECEDING_WHITESPACE: &str = "^([[:space:]])";
pub(super) const OPTIONAL_PRECEDING_WHITESPACE: &str = "^([[:space:]])*";
pub(super) const OPTIONAL_EXCLAMATION: &str = "(!)?";
pub(super) const ANGULAR_TYPE: &str =
    "(build|chore|ci|docs|feat|fix|perf|refactor|revert|style|test)";
pub(super) const EMPTY_SCOPE: &str = r"\(([[:space:]])*\)";
pub(super) const TYPE: &str = r"([[:alpha:]])+";
pub(super) const OPTIONAL_SCOPE: &str = r"(\([[:alpha:]]+(-[[:alpha:]]+)*\))?";
pub(super) const EMPTY_SCOPE_OR_SCOPE: &str = r"(\(.*\))";
pub(super) const OPTIONAL_EMPTY_SCOPE_OR_SCOPE: &str = r"(\(.*\))?";

lazy_static! {
    pub(super) static ref IGNORE_TYPE_AND_SCOPE_LINTING_ERRORS: String = format!(
        "{OPTIONAL_PRECEDING_WHITESPACE}{TYPE}{OPTIONAL_EXCLAMATION}{OPTIONAL_EMPTY_SCOPE_OR_SCOPE}{OPTIONAL_EXCLAMATION}:",
    );
}
