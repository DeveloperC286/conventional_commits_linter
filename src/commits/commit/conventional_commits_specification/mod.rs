use super::*;

pub(super) mod empty_scope;
pub(super) mod exclamation_mark_before_scope;
pub mod message_length;
pub(super) mod no_description_after_type_and_scope;
pub(super) mod no_space_after_colon_preceding_type_and_scope;
pub(super) mod preceding_whitespace;

static CONVENTIONAL_COMMITS_REGEX: OnceLock<Regex> = OnceLock::new();

pub(super) fn lint(commit_message: &str) -> Result<(), LintingError> {
    let regex = CONVENTIONAL_COMMITS_REGEX.get_or_init(|| {
        Regex::new(&format!(
            r"^{TYPE}{OPTIONAL_SCOPE}{OPTIONAL_EXCLAMATION}: [^[[:space:]]]+"
        ))
        .unwrap()
    });

    match regex.is_match(commit_message) {
        true => Ok(()),
        false => Err(LintingError::NonConventionalCommitsSpecification),
    }
}
