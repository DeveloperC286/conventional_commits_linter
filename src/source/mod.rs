#[derive(Clone, Copy, PartialEq, Eq)]
pub(crate) enum Source {
    Git,
    CommitMessage,
}
