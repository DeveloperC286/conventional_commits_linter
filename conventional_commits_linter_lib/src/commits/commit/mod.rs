use crate::LintingError;

mod allow_angular_type_only;
mod constants;
mod conventional_commits_specification;

use self::constants::*;

#[derive(Clone, PartialEq, Eq, Hash)]
pub(crate) struct Commit {
    pub(crate) hash: Option<String>,
    pub(crate) message: String,
}

impl Commit {
    pub(crate) fn from_commit_message<T: Into<String>>(message: T) -> Commit {
        Commit {
            hash: None,
            message: message.into(),
        }
    }

    pub(crate) fn from_git(commit: &git2::Commit) -> Commit {
        let message = match commit.message().map(|m| m.to_string()) {
            Some(message) => {
                trace!(
                    "Found the commit message {message:?} for the commit with the hash '{}'.",
                    commit.id()
                );

                message
            }
            None => {
                warn!(
                    "Can not find commit message for the commit with the hash '{}'.",
                    commit.id()
                );

                String::new()
            }
        };

        Commit {
            hash: Some(commit.id().to_string()),
            message,
        }
    }

    pub(crate) fn lint(&self, allow_angular_type_only: bool) -> Vec<LintingError> {
        let mut linting_errors = vec![];

        match conventional_commits_specification::lint(&self.message) {
            Ok(()) => {}
            Err(linting_error) => {
                linting_errors.push(linting_error);

                match conventional_commits_specification::preceding_whitespace::lint(&self.message)
                {
                    Ok(()) => {}
                    Err(linting_error) => {
                        linting_errors.push(linting_error);
                    }
                }

                match conventional_commits_specification::empty_scope::lint(&self.message) {
                    Ok(()) => {}
                    Err(linting_error) => {
                        linting_errors.push(linting_error);
                    }
                }

                match conventional_commits_specification::exclamation_mark_before_scope::lint(
                    &self.message,
                ) {
                    Ok(()) => {}
                    Err(linting_error) => {
                        linting_errors.push(linting_error);
                    }
                }

                match conventional_commits_specification::no_space_after_colon_preceding_type_and_scope::lint(&self.message) {
                Ok(()) => {}
                Err(linting_error) => {
                    linting_errors.push(linting_error);
                }
            }

                match conventional_commits_specification::no_description_after_type_and_scope::lint(
                    &self.message,
                ) {
                    Ok(()) => {}
                    Err(linting_error) => {
                        linting_errors.push(linting_error);
                    }
                }
            }
        }

        if allow_angular_type_only {
            match allow_angular_type_only::lint(&self.message) {
                Ok(()) => {}
                Err(linting_error) => {
                    linting_errors.push(linting_error);
                }
            }
        }

        linting_errors
    }
}

#[cfg(test)]
mod tests;
