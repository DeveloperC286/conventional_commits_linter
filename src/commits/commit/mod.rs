use std::sync::OnceLock;

use regex::Regex;

use crate::commit_type::CommitType;
use crate::linting_error::LintingError;

mod allow_angular_type_only;
mod constants;
mod conventional_commits_specification;

use self::constants::*;

#[derive(Clone, PartialEq, Eq, Hash)]
pub(crate) struct Commit {
    pub(crate) hash: Option<String>,
    pub(crate) message: String,
}

impl Commit {
    pub(crate) fn from_commit_message<T: Into<String>>(message: T) -> Commit {
        Commit {
            hash: None,
            message: message.into(),
        }
    }

    pub(crate) fn from_git(commit: &git2::Commit) -> Commit {
        let message = match commit.message().map(|m| m.to_string()) {
            Some(message) => {
                trace!(
                    "Found the commit message {message:?} for the commit with the hash '{}'.",
                    commit.id()
                );

                message
            }
            None => {
                warn!(
                    "Can not find commit message for the commit with the hash '{}'.",
                    commit.id()
                );

                String::new()
            }
        };

        Commit {
            hash: Some(commit.id().to_string()),
            message,
        }
    }

    pub(crate) fn lint(
        &self,
        commit_type: &CommitType,
        max_commit_title_length: usize,
        lowercase_scope: bool,
    ) -> Vec<LintingError> {
        info!("Linting the commit message {:?}.", self.message);
        let mut linting_errors = vec![];

        match conventional_commits_specification::lint(&self.message) {
            Ok(()) => {}
            Err(linting_error) => {
                linting_errors.push(linting_error);

                match conventional_commits_specification::preceding_whitespace::lint(&self.message)
                {
                    Ok(()) => {}
                    Err(linting_error) => {
                        linting_errors.push(linting_error);
                    }
                }

                match conventional_commits_specification::empty_scope::lint(&self.message) {
                    Ok(()) => {}
                    Err(linting_error) => {
                        linting_errors.push(linting_error);
                    }
                }

                match conventional_commits_specification::exclamation_mark_before_scope::lint(
                    &self.message,
                ) {
                    Ok(()) => {}
                    Err(linting_error) => {
                        linting_errors.push(linting_error);
                    }
                }

                match conventional_commits_specification::no_space_after_colon_preceding_type_and_scope::lint(&self.message) {
                Ok(()) => {}
                Err(linting_error) => {
                    linting_errors.push(linting_error);
                }
            }

                match conventional_commits_specification::no_description_after_type_and_scope::lint(
                    &self.message,
                ) {
                    Ok(()) => {}
                    Err(linting_error) => {
                        linting_errors.push(linting_error);
                    }
                }
            }
        }

        if *commit_type == CommitType::Angular {
            match allow_angular_type_only::lint(&self.message) {
                Ok(()) => {}
                Err(linting_error) => {
                    linting_errors.push(linting_error);
                }
            }
        }

        if lowercase_scope {
            match conventional_commits_specification::lowercase_scope::lint(&self.message) {
                Ok(()) => {}
                Err(linting_error) => {
                    linting_errors.push(linting_error);
                }
            }
        }

        // Check message length regardless of conventional commits compliance
        let max_length = if max_commit_title_length == 0 {
            None
        } else {
            Some(max_commit_title_length)
        };
        match conventional_commits_specification::message_length::lint(&self.message, max_length) {
            Ok(()) => {}
            Err(linting_error) => {
                linting_errors.push(linting_error);
            }
        }

        linting_errors
    }
}

#[cfg(test)]
mod tests;
