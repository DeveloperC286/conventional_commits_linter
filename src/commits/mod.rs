use std::collections::{HashMap, VecDeque};

use anyhow::{bail, Context, Result};
use git2::{Oid, Repository, Revwalk};

use crate::commit_type::CommitType;
use crate::commits::commit::Commit;
use crate::history_mode::HistoryMode;
use crate::linting_error::LintingError;
use crate::linting_errors::LintingErrors;
use crate::source::Source;

pub mod commit;

#[cfg(not(test))]
pub struct Commits {
    commits: VecDeque<Commit>,
    source: Source,
}

#[cfg(test)]
pub struct Commits {
    pub(crate) commits: VecDeque<Commit>,
    pub(crate) source: Source,
}

impl Commits {
    pub fn from_commit_message<T: Into<String>>(commit_message: T) -> Commits {
        Commits {
            commits: VecDeque::from(vec![Commit::from_commit_message(commit_message)]),
            source: Source::CommitMessage,
        }
    }

    pub fn from_git<T: AsRef<str>>(
        repository: &Repository,
        git: T,
        history_mode: HistoryMode,
    ) -> Result<Commits> {
        let oid = parse_to_oid(repository, git.as_ref()).or_else(|error| {
            get_reference_oid(repository, git.as_ref()).map_err(|e| error.context(e))
        })?;

        let commits = get_commits_till_head_from_oid(repository, oid, history_mode)?;
        Ok(Commits {
            commits,
            source: Source::Git,
        })
    }

    pub fn lint(
        self,
        commit_type: &CommitType,
        max_commit_title_length: usize,
    ) -> Option<LintingErrors> {
        let mut errors: HashMap<Commit, Vec<LintingError>> = HashMap::new();

        for commit in self.commits.iter().cloned() {
            let commit_errors = commit.lint(commit_type, max_commit_title_length);

            if !commit_errors.is_empty() {
                warn!(
                    "Found {} linting errors for the commit {:?}.",
                    commit_errors.len(),
                    commit.hash
                );
                errors.insert(commit, commit_errors);
            }
        }

        if errors.is_empty() {
            None
        } else {
            Some(LintingErrors::from(self.source, self.commits, errors))
        }
    }
}

fn get_commits_till_head_from_oid(
    repository: &Repository,
    from_commit_hash: Oid,
    history_mode: HistoryMode,
) -> Result<VecDeque<Commit>> {
    fn get_revwalker(
        repository: &Repository,
        from_commit_hash: Oid,
        history_mode: HistoryMode,
    ) -> Result<Revwalk<'_>> {
        let mut commits = repository.revwalk()?;
        if history_mode == HistoryMode::First {
            commits.simplify_first_parent()?;
        }
        commits.push_head()?;

        commits.hide(from_commit_hash).context(format!(
            "Can not find a commit with the hash '{from_commit_hash}'."
        ))?;
        Ok(commits)
    }

    let revwalker = get_revwalker(repository, from_commit_hash, history_mode)?;
    let mut commits = VecDeque::new();

    for oid in revwalker {
        let oid = oid?;
        let commit = repository.find_commit(oid)?;
        let commit = Commit::from_git(&commit);
        commits.push_front(commit);
    }

    if commits.is_empty() {
        bail!("No Git commits within the provided range.");
    }

    info!("Found {} commits within the provided range.", commits.len());
    Ok(commits)
}

fn get_reference_oid(repository: &Repository, matching: &str) -> Result<Oid> {
    let reference = repository
        .resolve_reference_from_short_name(matching)
        .context(format!(
            "Could not find a reference with the name {matching:?}."
        ))?;
    debug!(
        "Matched {:?} to the reference {:?}.",
        matching,
        reference.name().unwrap()
    );
    let commit = reference.peel_to_commit()?;
    Ok(commit.id())
}

fn parse_to_oid(repository: &Repository, oid: &str) -> Result<Oid> {
    match oid.len() {
        1..=39 => {
            debug!("Attempting to find a match for the short commit hash {oid:?}.");
            let matching_oid_lowercase = oid.to_lowercase();

            let mut revwalker = repository.revwalk()?;
            revwalker.push_head()?;

            let matched_commit_hashes: Vec<Oid> = revwalker
                .filter_map(|result| match result {
                    Ok(oid) => {
                        let oid_lowercase = oid.to_string().to_lowercase();

                        if oid_lowercase.starts_with(&matching_oid_lowercase) {
                            debug!("Found a match for the short commit hash {oid:?}.");
                            return Some(oid);
                        }

                        None
                    }
                    Err(_) => None,
                })
                .collect();

            match matched_commit_hashes.len() {
                0 => {
                    bail!(
                        "No commit hashes start with the provided short commit hash {:?}.",
                        matching_oid_lowercase
                    );
                }
                1 => Ok(*matched_commit_hashes.first().unwrap()),
                _ => {
                    bail!("Ambiguous short commit hash, the commit hashes {:?} all start with the provided short commit hash {:?}.", matched_commit_hashes, matching_oid_lowercase);
                }
            }
        }
        _ => git2::Oid::from_str(oid).context(format!("{oid:?} is not a valid commit hash.")),
    }
}
