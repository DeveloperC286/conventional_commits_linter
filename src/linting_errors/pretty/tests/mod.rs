use rstest::rstest;

use crate::commit_type::CommitType;
use crate::commits::commit::Commit;
use crate::commits::Commits;
use crate::source::Source;

const DEFAULT_COMMIT_TYPE: &CommitType = &CommitType::Any;

#[rstest(
    commit_message,
    snapshot_name,
    case(
        "feat()!: yargs-parser now throws on invalid combinations of config\n\n",
        "test_pretty_print_case_1"
    ),
    case("Release 1.0.126", "test_pretty_print_case_2"),
    case(
        "Merge branch 'fix-ocsp-signer-check' into 'master'",
        "test_pretty_print_case_3"
    )
)]
fn test_pretty_print_from_commit_message(commit_message: &str, snapshot_name: &str) {
    // Given
    let commits = crate::commits::Commits::from_commit_message(commit_message);
    let linting_errors = commits.lint(DEFAULT_COMMIT_TYPE).unwrap();

    // When
    let pretty_print = linting_errors.pretty();

    // Then
    insta::assert_snapshot!(snapshot_name, pretty_print);
}

#[rstest(
    commit_messages,
    snapshot_name,
    case(
        &["feat()!: yargs-parser now throws on invalid combinations of config\n\n", "doc(webpack):webpack example (#1436)"],
        "test_pretty_print_from_git_case_1"
    ),
    case(
        &["refactor(ts support)!: ship yargs.d.ts (#1671)", "feat!: drop support for EOL Node 8 (#1686)"],
        "test_pretty_print_from_git_case_2"
    ),
)]
fn test_pretty_print_from_git(commit_messages: &[&str], snapshot_name: &str) {
    // Given
    let commits: Commits = Commits {
        source: Source::Git,
        commits: commit_messages
            .iter()
            .enumerate()
            .map(|(index, commit_message)| Commit {
                hash: Some(
                    git2::Oid::from_str(&format!("{index}"))
                        .unwrap()
                        .to_string(),
                ),
                message: commit_message.to_string(),
            })
            .collect(),
    };

    let linting_errors = commits.lint(DEFAULT_COMMIT_TYPE).unwrap();

    // When
    let pretty_print = linting_errors.pretty();

    //Then
    insta::assert_snapshot!(snapshot_name, pretty_print);
}
