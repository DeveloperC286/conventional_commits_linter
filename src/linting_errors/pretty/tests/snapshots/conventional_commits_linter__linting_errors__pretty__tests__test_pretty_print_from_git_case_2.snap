---
source: src/linting_errors/pretty/tests/mod.rs
expression: pretty_print
---
[1;31mCommit Hash[0m - 0000000000000000000000000000000000000000
[1;31mMessage[0m - "refactor(ts support)!: ship yargs.d.ts (#1671)"
	[1;31mX[0m - Commit title does not comply with the Conventional Commits V1.0.0 specification.

[1;31mX[0m - Found 1 separate linting errors across 1 commits.
